provider/queue.go

```diff
@@ -0,0 +1,124 @@
+package provider
+
+import (
+	"errors"
+	"gx/ipfs/QmR8BauakNcBa3RbE4nbQu76PDiJgoQgz8AJdhJuiU4TAw/go-cid"
+	ds "gx/ipfs/Qmf4xQhNomPNhrtZc67qSnfJSjxjXs9LWvknJtSXwimPrM/go-datastore"
+	"gx/ipfs/Qmf4xQhNomPNhrtZc67qSnfJSjxjXs9LWvknJtSXwimPrM/go-datastore/query"
+	"math"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type Queue struct {
+	// used to differentiate queues in datastore
+	// e.g. provider vs reprovider
+	name string
+
+	tail uint64
+	head uint64
+
+	lock sync.Mutex
+	datastore ds.Datastore
+}
+
+func NewQueue(name string, datastore ds.Datastore) *Queue {
+	q := &Queue{
+		name: name,
+		head: math.MaxUint64,
+		lock: sync.Mutex{},
+		datastore: datastore,
+	}
+	q.prepare()
+	return q
+}
+
+func (q *Queue) Enqueue(cid cid.Cid) error {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	nextKey := q.queueKey(q.tail)
+
+	if err := q.datastore.Put(nextKey, cid.Bytes()); err != nil {
+		return err
+	}
+
+	q.tail++
+	return nil
+}
+
+func (q *Queue) Dequeue() (cid.Cid, error) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	if q.IsEmpty() {
+		// TODO figure out how IPFS folks are doing custom errors and make this comply
+		return cid.Undef, errors.New("queue is empty")
+	}
+
+	nextKey := q.queueKey(q.head)
+	value, err := q.datastore.Get(nextKey)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	key, err := cid.Parse(value)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	if err := q.datastore.Delete(nextKey); err != nil {
+		return cid.Undef, err
+	}
+
+	q.head++
+	return key, nil
+}
+
+func (q *Queue) IsEmpty() bool {
+	return q.Length() == 0
+}
+
+func (q *Queue) Length() uint64 {
+	return q.tail - q.head
+}
+
+func (q *Queue) prepare() error {
+	query := query.Query{Prefix: q.queuePrefix()}
+	results, err := q.datastore.Query(query)
+	if err != nil {
+		return err
+	}
+
+	// TODO: this is maybe dumb
+	for entry := range results.Next() {
+		keyId := strings.TrimPrefix(entry.Key, q.queuePrefix())
+		id, err := strconv.ParseUint(keyId, 10, 64)
+		if err != nil {
+			return err
+		}
+
+		if id < q.head {
+			q.head = id
+		}
+
+		if id > q.tail {
+			q.tail = id
+		}
+	}
+
+	if q.head == math.MaxUint64 {
```

Comment:
NB: Is prepare ever called outside the constructor? Asking cause it wasn't clear when I first looked at this how q.head became MaxUint64, which is necessary for the above loop to work, and then I realized it was set in the constructor. Seems like for clarity on your algorithm, setting head to MaxUint64 should happen shortly before the for loop.

Also, I might instead of calling this "prepare" I might make it a function call getHeadTailFromDatastore() (uint64, uint64), not have it mutate q.head and q.tail at all, and then just call it in constructor and use the values to initialize head and tail? I dunno.

Oh, also, I'm just curious, is there a case in which the queue on disk could become dis-contiguous -- i.e. head = 2, tail = 5 but there is no key written with key = 3 ? Just curious if that could happen (maybe due to the go routines in Provide) and whether that would result in anything bad.